### PR TITLE
PSVITA port and fix for audio decoder on 3DS port.

### PR DESCRIPTION
**Pull request by Rinnegatamante**
As the title says.
PSVITA port lacks only of OGG support (both sounds and musics).
Everything else is fully working.

### CODE REVIEW

**Comment by Ghabry** [comment]
For the pics I propose: Any other suggestions?

[pics.zip](https://github.com/EasyRPG/Player/files/402405/pics.zip)
image1 crushed
pngout

**Comment by Ghabry** [comment]
Great. Some BGM features are missing. Though to implement them you need a reference to the AudioDecoder.

The missing fade is noticable in the Ib video, the music should fade out when the light begins to flicker iirc.

Here are the important lines, copied from sdl_audio:

```cpp
BGM_Play
// You can use GetVolume in the BGM thread to get the current volume, it takes care of fades
audio_decoder->SetFade(0, volume, fadein);
audio_decoder->SetPitch(pitch); // pitch is commented out in the argument list
bgm_starttick = DisplayUi->GetTicks();

BGM_GetTicks
return audio_decoder->GetTicks();

BGM_PlayedOnce
return audio_decoder->GetLoopCount() > 0;

BGM_Volume
// Only a convenience function, you have to still set it for the hardware using something like
// sceAudioOutSetVolume(audio_decoder->GetVolume(), ....)
audio_decoder->SetVolume(volume);

BGM_SetPitch
audio_decoder->SetPitch(pitch);

BGM_Fade
audio_decoder->SetFade(audio_decoder->GetVolume(), 0, fade);
bgm_starttick = DisplayUi->GetTicks();

SE_Play
Also has pitch and volume

Update
// The Fade needs a time delta to work:
if (audio_decoder && bgm_starttick > 0) {
	int t = DisplayUi->GetTicks();
	audio_decoder->Update(t - bgm_starttick);
	bgm_starttick = t;
}
```

**Comment by Rinnegatamante** [comment]
PlayedOnce is already implemented. Working on the other things.

**Comment by Ghabry** [comment]
People who play Yume Nikki will need the buttons N1, N3, N5 and N9 btw. (SHIFT is already mapped) Are there some keys on the controller available to map them? Maybe abuse one of the analog sticks.

There are also much more buttons, but because everybody plays Yume Nikki these are enough ;)

**Comment by Rinnegatamante** [comment]
Mhhh maybe we can map the retrotouch or the touchscreen. OR the right analog. Cause there aren't other buttons available.

**Comment by Rinnegatamante** [comment]
Little note: retrotouch and touchscreen = PSVITA TV not compatible.

**Comment by Ghabry** [comment]
The right analog offers at least 4 directions, so this sounds like a plan.

**Comment by Ghabry** [comment]
Oh and you have to set the paths where the Runtime Package is searched.

This is in this function:
https://github.com/EasyRPG/Player/blob/master/src/filefinder.cpp#L388

Default is
```
add_rtp_path("/data/rtp/" + version_str + "/");
```
If you think this path is fine just keep it and don't modify the function :)

This is also not altered for the 3DS version, though not sure if that's a big deal.

File: src/psp2_ui.cpp [hunk 1]
```diff
@@ -134,6 +138,18 @@ void Psp2Ui::ProcessEvents() {
 	trigger_state = (input.buttons & SCE_CTRL_RTRIGGER);
 	if ((trigger_state != old_state) && trigger_state) zoom_state = ((zoom_state + 1) % 3);
 	
+	// Left analog support
+	if (input.ly > 170) keys[Input::Keys::UP] = true;
+	else if (input.ly < 50) keys[Input::Keys::DOWN] = true;
+	else if (input.lx > 170) keys[Input::Keys::RIGHT] = true;
+	else if (input.lx < 50) keys[Input::Keys::LEFT] = true;
+
```
Comment (carstene1ns) [inline src/psp2_ui.cpp]: As this is technically a analog stick, you should define for VITA

```c++
#  define SUPPORT_JOYSTICK
#  define SUPPORT_JOYSTICK_AXIS
```
in `system.h` and use something like that:
```c++
	keys[Input::Keys::JOY_AXIS_X_LEFT] = (input.lx < 50);
	keys[Input::Keys::JOY_AXIS_X_RIGHT] = (input.lx > 170);
	keys[Input::Keys::JOY_AXIS_Y_DOWN] = (input.ly < 50);
	keys[Input::Keys::JOY_AXIS_Y_UP] = (input.ly > 170);
```

Comment (Rinnegatamante) [inline src/psp2_ui.cpp]: The same apply to 3DS port then with Circle Stick i suppose

**Comment by Rinnegatamante** [comment]
Talking back to this i previously said:
"Little note: retrotouch and touchscreen = PSVITA TV not compatible."

I discovered it's wrong since even on PSVITA TV there's a setting to emulate touchpad and touchscreen with analogs (basically pressing analogs + moving = emulating touch).
Looks still pretty uncomfortable to use but can be used if we need two extra buttons (just touchscreen tap and retrotouch tap).

**Comment by Ghabry** [comment]
Jenkins: Test this please

**Comment by Ghabry** [comment]
Jenkins: Test this please

File: builds/psvita/Makefile [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		
+PREFIX  = arm-vita-eabi
+CC      = $(PREFIX)-gcc
+CXX      = $(PREFIX)-g++
+CFLAGS  = -Wl,-q -O3 -DPSP2 -DHAVE_LIBSPEEXDSP \
+		-DSUPPORT_AUDIO -DHAVE_LIBSNDFILE \
+		-DHAVE_OGGVORBIS -DHAVE_MPG123 -DWANT_FMMIDI=1 \
+		-I$(VITASDK)/arm-vita-eabi/include/liblcf \
+		-I$(VITASDK)/arm-vita-eabi/include/pixman-1
+CXXFLAGS  = $(CFLAGS) -fno-exceptions -std=gnu++11
```
Comment (Ghabry) [inline builds/psvita/Makefile]: For better cooperation with our PR builder I ask you to add some lines here (they won't interfere with yours so is fine ;))

```
Before PREFIX:

ifeq ($(strip $(LIBLCF_DIR)),)
LIBLCF_DIR := $(TOOLCHAIN_DIR)
endif

To CFLAGS add:
		-I$(TOOLCHAIN_DIR)/include \
		-I$(LIBLCF_DIR)/include/liblcf \
		-I$(TOOLCHAIN_DIR)/include/pixman-1 \
		-L$(TOOLCHAIN_DIR)/lib \
		-L$(LIBLCF_DIR)/lib

To CXXFLAGS add (should make the file a bit smaller, we don't need RTTI):
		-fno-rtti

**Comment by Ghabry** [comment]
After adding the mentioned makefile lines, fixing the whitespacing in psp2_ui.cpp (some lines use spaces instead of tabs) and doing a rebase on master to get rid of the merge commits I think this one here is okay in it's current state.

```
(make a backup of your before if you are afraid)

git rebase upstream/master
git push -f origin
```

As a long term solution I think we should add another audio system that mixes BGM+SE together in software so the porter only has to play that buffer, also works around the channel limits some hardware has. Will make the life much easier for all non-SDL ports :+1:

File: builds/psvita/Makefile [hunk 3]
```diff
@@ -32,10 +32,10 @@ CFLAGS  = -Wl,-q -O3 -DPSP2 -DHAVE_LIBSPEEXDSP \
 		-I$(VITASDK)/arm-vita-eabi/include/liblcf \
 		-I$(VITASDK)/arm-vita-eabi/include/pixman-1
```
Comment (Ghabry) [inline builds/psvita/Makefile]: This line needs a "\" at the end.
When Github Desktop support "ammend commit" you can tick that checkbox to integrate it into this commit.

**Comment by Ghabry** [comment]
Jenkins: Test this please

File: src/filefinder.cpp [hunk 4]
```diff
@@ -36,10 +36,16 @@
...
 #  endif
 #else
-#  include <dirent.h>
+#  ifdef PSP2
+#    include <psp2/io/dirent.h>
+#    include <psp2/io/stat.h>
+#    define S_ISDIR SCE_S_ISDIR
+#  else
+#    include <dirent.h>
+#	 include <sys/stat.h>
```
Comment (carstene1ns) [inline src/filefinder.cpp]: Found a stray `<Tab>` here.

File: src/filefinder.cpp [hunk 5]
```diff
@@ -675,29 +687,52 @@ FileFinder::Directory FileFinder::GetDirectoryMembers(const std::string& path, F
 #  define readdir _wreaddir
 #elif _3DS
 	std::string wpath = path + "/";
+#elif PSP2
+#  define opendir sceIoDopen
+#  define closedir sceIoDclose
+#  define wpath path
+#  define dirent SceIoDirent
+#  define readdir sceIoDread
 #else
```
Comment (carstene1ns) [inline src/filefinder.cpp]: Question: Why the need to define these and then later use the redefinitions?
Would make more sense to use them directly to me.

File: src/main_data.cpp [hunk 6]
```diff
@@ -107,6 +109,7 @@ void Main_Data::Init() {
 					FS_Path filePath2=fsMakePath(PATH_ASCII, &mainDir[5]);
 					FSUSER_CreateDirectory(archive,filePath2, FS_ATTRIBUTE_DIRECTORY);
 					FSUSER_CloseArchive(archive);
+
```
Comment (carstene1ns) [inline src/main_data.cpp]: Stray empty line.

File: src/psp2_decoder.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _EASYRPG_3DS_DECODER_H_
+#define _EASYRPG_3DS_DECODER_H_
```
Comment (carstene1ns) [inline src/psp2_decoder.h]: This is likely wrong from copying it over. ;-)

File: src/psp2_ui.h [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _EASYRPG_3DSUI_H_
+#define _EASYRPG_3DSUI_H_
```
Comment (carstene1ns) [inline src/psp2_ui.h]: Same applies here.

File: builds/psvita/Makefile [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+all: $(TARGET).vpk
+
+$(TARGET).vpk: $(TARGET).velf
+	vita-make-fself $< .\release\eboot.bin
+	vita-mksfoex -s TITLE_ID=$(APP_ID) "$(APP_TITLE)" param.sfo
+	cp -f param.sfo ./release/sce_sys/param.sfo
+	
+	#------------ Comment this if you don't have 7zip ------------------
+	7za a -tzip ./$(TARGET).vpk -r .\release\sce_sys\* .\release\eboot.bin
```
Comment (Ghabry) [inline builds/psvita/Makefile]: Please convert the \ to /, will also work on Windows

**Comment by carstene1ns** [comment]
This is my working Makefile rule:
```make
$(TARGET).vpk: $(TARGET).velf
	vita-make-fself $< release/eboot.bin
	vita-mksfoex -s TITLE_ID=$(APP_ID) "$(APP_TITLE)" release/sce_sys/param.sfo
	7za a -tzip $(TARGET).vpk ./release/sce_sys ./release/eboot.bin
```

File: src/psp2_ui.cpp [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+	keys[Input::Keys::JOY_AXIS_X_LEFT] = (input.lx < 50);
+	keys[Input::Keys::JOY_AXIS_X_RIGHT] = (input.lx > 170);
+	keys[Input::Keys::JOY_AXIS_Y_DOWN] = (input.ly > 170);
+	keys[Input::Keys::JOY_AXIS_Y_UP] = (input.ly < 50);
+	
+	// Right analog support for extra buttons
+	if (input.ry > 170) keys[Input::Keys::N1] = true;
+	else if (input.ry < 50) keys[Input::Keys::N3] = true;
+	else if (input.rx > 170) keys[Input::Keys::N5] = true;
+	else if (input.rx < 50) keys[Input::Keys::N9] = true;
```
Comment (Ghabry) [inline src/psp2_ui.cpp]: A spanish user on twitter reported that the keys of the right joystick get stuck in "on" state when used.
Looks like he is rìght :)

**Comment by Ghabry** [comment]
```
HAVE_ZIP := $(shell command -v zip 2> /dev/null)

all: $(TARGET).vpk

$(TARGET).vpk: $(TARGET).velf
        vita-make-fself $< release/eboot.bin
        vita-mksfoex -s TITLE_ID=$(APP_ID) "$(APP_TITLE)" release/sce_sys/param.sfo
ifdef HAVE_ZIP
        cd release; zip ../$(TARGET).vpk -r ./sce_sys ./eboot.bin
else
        7z a -tzip ./$(TARGET).vpk -r ./release/sce_sys/* ./release/eboot.bin
endif
```

Another solution that could make both happy (Replace space with tabs!). The 7z line is wrong, use your Windows one there with the "\"

**Comment by Rinnegatamante** [comment]
On my side it apparently works fine so i suppose it's ready for a push

**Comment by Ghabry** [comment]
Jenkins: Test this please

**Pull request merged**